### PR TITLE
feat(SLA): Apply SLA to any document

### PR DESCRIPTION
**Pull request by hrwX**
This PR enables setting Service Level Agreement for any doctype.

### Document Type field added to SLA doctype

document_type

### SLA fields

On creating the SLA, an SLA section will get added to the selected doctype with the following fields (custom fields for standard doctype and docfields for custom doctype).
  - Service Level Agreement
  - Priority
  - Response By
  - Response By Variance
  - Resolution By
  - Resolution By Variance
  - First Responded On
  - On Hold Since
  - Total Hold Time
  - Agreement Status
  - SLA creation
  - Resolution Date

custom_dt_with_sla

### Hold Time

Hold Time is captured if the **Pause SLA On**  configuration is setup in SLA

https://user-images.githubusercontent.com/24353136/120075658-d9c84780-c0bf-11eb-9fa5-f17a9fb0dc8c.mp4

### SLA Fulfilled On

Allows you to configure statuses for SLA fulfillment

image

### Apply SLA for Resolution Time

For a few instances like applying SLA on leads, there is no resolution SLA required. The only metric that matters is the First Response Time. Hence, the SLA application for resolution time is now optional. If it is disabled, the SLA status is validated as Fulfilled / Failed on the basis of first response time only.

image

https://user-images.githubusercontent.com/24353136/120075975-22343500-c0c1-11eb-9dd8-4e76478e540c.mp4

### Hooks
   1. Applying SLA
   2. Setting hold time on communication update
   3. Setting SLA variance

Documentation PR: https://github.com/frappe/erpnext_documentation/pull/340

### CODE REVIEW

**Comment by ruchamahabal** [comment]
@hrwx Why have we removed the Employee Group field from SLA doctype in this PR? Any specific reason?

File: erpnext/support/doctype/service_level_agreement/service_level_agreement.py [hunk 1]
```diff
@@ -101,78 +105,169 @@ def get_service_level_agreement_priority(self, priority):
...
+
+		service_level_agreement_fields = get_service_level_agreement_fields()
+		meta = frappe.get_meta(self.document_type, cached=False)
+
+		if meta.custom:
+			self.create_docfields(meta, service_level_agreement_fields)
+		else:
+			self.create_custom_fields(meta, service_level_agreement_fields)
+
+	def on_trash(self):
```
Comment (nextchamp-saqib) [inline erpnext/support/doctype/service_level_agreement/service_level_agreement.py:121]: Can the custom fields be deleted on trash?

Comment (ruchamahabal) [reply erpnext/support/doctype/service_level_agreement/service_level_agreement.py:121]: Unsure. IMO, deleting fields is a bit risky. If the users have some custom app or have created custom fields by themselves, they wouldn't want them to be deleted.

**Comment by barredterra** [comment]
@hrwX do you remember why you decided to convert timestamps to the user timezone in the backend for this feature? It seems like all other code works with the system timezone in the backend and lets the UI/frontend convert to/from the user timezone.

We're currently [contemplating to remove this](https://github.com/frappe/frappe/pull/19933#issuecomment-1455131885).

**Pull request merged**